osidb/mixins.py

```diff
@@ -597,26 +603,91 @@ def set_public_nested(self):
         Change internal ACLs to public ACLs for all related Flaw objects and save them.
         The only exception is "snippets", which should always have internal ACLs.
         The Flaw itself will be saved later to avoid duplicate operations.
+
+        This method collects all objects that need to be updated and uses bulk_update
+        to minimize database queries.
+        """
+
+        # Collect all objects that need to be updated (using dict to avoid duplicates)
+        # Key: (model_class, pk), Value: object instance
+        objects_to_update = {}
+        # Track visited objects to prevent infinite recursion in circular relationships
+        visited = set()
+        self._collect_objects_for_public_update(objects_to_update, visited)
+
+        # Get list of unique objects
+        valid_objects = list(objects_to_update.values())
+
+        if not valid_objects:
+            return
+
+        # Set updated_dt on all objects before bulk update
+        # Cut off microseconds to match TrackingMixin.save() behavior
+        now = timezone.now().replace(microsecond=0)
+        for obj in valid_objects:
+            if hasattr(obj, "updated_dt"):
+                obj.updated_dt = now
+
+        # Group objects by model type for bulk_update
+        from collections import defaultdict
+
+        objects_by_model = defaultdict(list)
+        for obj in valid_objects:
+            objects_by_model[type(obj)].append(obj)
+
+        # Bulk update each model type
+        # Fields to update: acl_read, acl_write, and updated_dt
+        for model_class, objects in objects_by_model.items():
+            if objects:
+                # Determine fields for this model type
+                model_update_fields = ["acl_read", "acl_write"]
+                if hasattr(model_class, "updated_dt"):
+                    model_update_fields.append("updated_dt")
+
+                model_class.objects.bulk_update(
+                    objects,
+                    model_update_fields,
+                    batch_size=100,  # Process in batches to avoid memory issues
+                )
+
+                # Update audit history for all updated objects
+                # Each object needs its history set to public after ACL update
+                for obj in objects:
+                    if hasattr(obj, "set_history_public"):
+                        obj.set_history_public()
+
+    def _collect_objects_for_public_update(self, objects_to_update, visited):
+        """
+        Recursively collect all related objects that need to have their ACLs
+        updated to public. The Flaw itself is not collected as it's saved separately.
+
+        Uses a dictionary keyed by (type, pk) to track seen objects and avoid duplicates.
+        Uses a set to track visited objects to prevent infinite recursion.
         """
         from osidb.models import Flaw
 
+        # Create a unique key for this object to track if we've visited it
```

Comment:
nitpick: The following line does not create anything. I would more naturally put the comment three lines below.